#15 Add navbar

chelseaharris137 opened the pull request:
## Description of Proposed Changes

This PR resolves #15 

### Type of Change

- [x] New feature (non-breaking change which adds functionality)

## Testing

Spin up the application. The navbar should appear on the homepage. Test mobile responsive in devtools.

## Checklists

### GitHub

- [x] Related issue appears at beginning of pull request title with pound sign, and title describes the changes being proposed.
- [x] Related issue is linked in pull request description using [closing keywords](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword).
- [x] The appropriate label has been chosen for this pull request.
- [x] The correct project has been selected for this pull request.
- [x] All commits in this branch, including merge commits, begin with the issue number and a pound sign.

### Code Standards

- [x] Code follows the style guidelines of this project.
- [x] Code comments have been added, particularly in hard-to-understand areas.
- [x] I have performed a self-review of my code.
- [x] Any dependent changes have been merged and published in downstream modules.
- [x] These changes generate no new warnings.

### Documentation and Testing

- [x] New and existing unit tests pass locally with these changes.
- [x] Corresponding changes to the documentation have been made.
- [x] Tests have been added, where appropriate; or, no tests are relevant for this pull request.
- [x] This pull request contains UI changes, and screenshots of those UI images appear below; or, this pull request contains no UI changes.

## Screenshots

![health-tracker-light-1](https://github.com/Rainbow-Devs/health-tracker-frontend/assets/4006420/84ca0943-34b4-4f46-a876-317da3436a41)
![health-tracker-light-2](https://github.com/Rainbow-Devs/health-tracker-frontend/assets/4006420/9348a494-4959-4637-bab6-8508e8e7c086)
![health-tracker-dark-1](https://github.com/Rainbow-Devs/health-tracker-frontend/assets/4006420/ddf604a9-1086-4cec-840d-3596fe81bf5b)
![health-tracker-dark-2](https://github.com/Rainbow-Devs/health-tracker-frontend/assets/4006420/5cd0fff8-1a85-459e-8801-9a16d4baa478)

chelseaharris137 committed:
#15 Add navbar

chelseaharris137 commented on src/styles/classes/fontClasses.tsx:
```diff
@@ -0,0 +1,7 @@
+export const fontSizeClasses: { [key: string]: string } = {
+  small: "text-sm"
+};
+
+export const fontWeightClasses: { [key: string]: string } = {
+  medium: "font-medium"
+};
```
I've taken this object-oriented approach to avoid string concatenation while allowing dynamic classes. I'm new to Tailwind, so I am open to alternate solutions here/advice on best implementing re-usable components & atoms.

seidior replied:
We'll have to chat off GitHub about this but I'd like to learn more about your strategy here before proceeding with this approach.

chelseaharris137 commented on src/components/Navbar.tsx:
```diff
@@ -0,0 +1,57 @@
+"use client";
+import { Button, CustomLink } from "@/atoms";
+import { useRouter } from "next/navigation";
+import Image from "next/image";
+
+export const Navbar = () => {
+  const router = useRouter();
+
+  const navigateOnClick = (path: string) => router.push(path);
+
```
I think we will eventually want to pass this navigation function down as a prop but for now leaving it here.

chelseaharris137 commented:
I might have missed this at some point -> what is our UI testing strategy? I have not implemented tests but can update once I have more info.

CodeDiversity commented:
Overall, I think it looks amazing. Nice job. I know we didn't have mobile comps, but I am thinking that a hamburger menu might better suit us on mobile. Thoughts? If not hamburger menu maybe make them full width links/actions with borders on top and bottom only. Totally up to you.

![image](https://github.com/Rainbow-Devs/health-tracker-frontend/assets/40902786/fda1f3ca-53b2-47da-a91f-d47681f378b8)


![image](https://github.com/Rainbow-Devs/health-tracker-frontend/assets/40902786/076ce9a4-22df-4973-bec5-b14bb74ae105)

seidior commented:
> I might have missed this at some point -> what is our UI testing strategy? I have not implemented tests but can update once I have more info.

We have not decided on test libraries for frontend or backend code. I have experience with some of the most popular ones, but I'd love to hear from the community what works best for them before guiding toward one library over another.

## Resources:

- [Next.js: Testing](https://nextjs.org/docs/pages/building-your-application/optimizing/testing)

seidior commented on src/atoms/Button.tsx:
```diff
@@ -0,0 +1,35 @@
+import React from "react";
```
React should already be in scope for all `.jsx` and `.tsx` files, so this line is no longer necessary.

seidior commented on tailwind.config.cjs:
```diff
@@ -3,16 +3,23 @@ module.exports = {
         "alert-red": "#BC1C21",
         "dark": "#090E34",
-        "light-gray": "#637381",
+        "dark-gray": "#637381",
```
You can do [gradations of colors](https://tailwindcss.com/docs/theme#colors) so that you can use more-standard Tailwind classes like `text-gray-500`, etc.

chelseaharris137 replied:
Updated

seidior commented on src/atoms/CustomLink.tsx:
If we're thinking of Atomic Design here, we're probably going to structure these things differently:

- Organism: NavBar
  - Atom: NavBarLogo
  - Molecule: NavBarActions
    - Atom: NavBarActionButton (variant: LinkStyle)
    - Atom: NavBarActionButton (variant: ButtonStyle)

Under the surface, both of these should be buttons, since they are actions to be performed rather than navigational items. (In theory, they should be showing a modal rather than navigating to a page.) Additionally, `<button>`s are tab-navigable by default, where `<a>` tags are not, and we've stated that accessibility is one of our goals.

seidior commented on src/app/page.tsx:
```diff
@@ -1,113 +1,81 @@
-        </p>
-        <div className="fixed bottom-0 left-0 flex h-48 w-full items-end justify-center bg-gradient-to-t from-white via-white dark:from-black dark:via-black lg:static lg:h-auto lg:w-auto lg:bg-none">
+    <>
+      <Navbar />
```
This should go in `layout.tsx` instead of `page.tsx`, since the former is shared between all pages and the latter is just supposed to be what is unique about each page.

seidior commented on src/app/page.tsx:
```diff
@@ -1,113 +1,81 @@
+    <>
+      <Navbar />
+
+      <main className="flex min-h-screen flex-col items-center justify-between bg-white p-24">
```
I think at this point it'd be good to remove all the boilerplate content.

seidior commented on tailwind.config.cjs:
```diff
@@ -3,16 +3,23 @@ module.exports = {
   content: [
     "./src/pages/**/*.{js,ts,jsx,tsx,mdx}",
```
If for whatever reason we are going to stick with the `src` directory approach, all of these lines can just be simplified to 

```javascript
"./src/**/*.{js,ts,jsx,tsx,mdx}"
```

seidior commented on tailwind.config.cjs:
```diff
@@ -3,16 +3,23 @@ module.exports = {
   theme: {
     extend: {
+      fontSize: {
+        sm: '16px'
```
We already have a '16px' text size and that's `text-base`. What is the reasoning or motivation behind redefining `text-sm` to be '16px' (1 rem) instead of its standard '14px' (0.875 rem)?

seidior reviewed (changes requested):
Thank you for taking the lead on this, but I think jumping into this ticket before addressing the Atomic Design ticket is a little bit premature at this stage, since we should probably set a good foundation for us going forward.

I like the trick you used here with the `index.tsx` files in the directories to make it easier to incorporate elements on each of these pages. This may be a premature optimization speaking but I'm not sure this is the best strategy for performance reasons, unfortunately.

Let's touch base Monday before proceeding further on this ticket so that everyone's on the same page. I've given you some preliminary feedback here but it's far from an exhaustive list of things to address.

It's really great to start to have momentum on building out the app. It's so exciting! Thanks again for taking the charge and getting us started.

seidior committed:
#15 Merge latest from main

seidior committed:
#15 Add cva, upgrade packages

seidior committed:
#15 Add additional dependencies

seidior committed:
#15 Atomic Design Header

Side benefit: using Next.js parallel and intercepting routes to have a modal dialog and login page in one.

seidior commented:
An update on this PR:

- In order to save time with a back-and-forth over GitHub, I opted to address some of the issues from the pull request review in a separate branch, giving the room to do some research to establish baseline standards for developers contributing Atomic Design elements in the future while preserving as much of the great work that was already done in the ticket, then chat with Chelsea directly about the changes.
- Chelsea gave the go-ahead to merge the changes into the branch for review and for additional work before the next review:
  - Another pass will be given to the color changes as well as any other modifications necessary to find a compromise between what is designed and what will give the best developer experience for other contributors in the future.
  - More consistency will be added in the styles, such as between background and outline colors as well as consistency with Tailwind conventions (while preserving the AriaKit styles so as not to break these components).
  - Any changes to the design will be communicated with the design team for their approval and sign-off.
  - Finally, the screenshots in the PR above will be updated with the final design for this.

After Chelsea is done improving my contributed changes, I'll take another look and flip my review to "approve" so I'm not a roadblock on this anymore :)

chelseaharris137 committed:
#15 Add gray color gradations

chelseaharris137 committed:
#15 Update dark mode design

chelseaharris137 commented:
Hey All - I've made some updates to dark mode and have reached out to Wren to discuss flexibility with our themeing colors. Waiting to hear back from her. I'm going to have to step away from this for now but feel free to make any additional updates needed.

seidior commented:
Reviewing the PR right now, and this is the dark mode display for me:

![dark-mode-monday-morning](https://github.com/Rainbow-Devs/health-tracker-frontend/assets/4006420/ae1249ae-b0e3-460d-8992-10bb88fa32cf)

Given that Chelsea reminded us about unavailability, I'll go ahead and make the changes to improve the display here as well as update the screenshots in the pull request description when those changes have been pushed. Thanks for your work on the ticket!

seidior committed:
#15 Design refinements

Fixed text display issue in dark mode, expanded the gray color palette to include more stops, and refined the design to better adapt to the theme colors while preserving accessibility as much as possible.

seidior committed:
#15 Minor package updates

seidior reviewed (approved):
Switching to a conditional approval for my review on this ticket. I'd love to have the design team's input before proceeding.

seidior committed:
#15 Change NavBar light mode background

seidior commented:
Very slight tweak based on feedback from the design team: the NavBar is now translucent white instead of translucent gray, which better matches the design and gives us better scores on accessibility.

<img width="1024" alt="new-light" src="https://github.com/Rainbow-Devs/health-tracker-frontend/assets/4006420/47067d2f-b3b3-4790-8d5f-4aecdce66f70">

Other than that, this is now approved for merger.
